### PR TITLE
Add SPI prepare and SPI exec_plan for R client

### PR DESCRIPTION
**Pull request by HaozhouWang**
1. R clients now support pg.spi.prepare and pg.spi.execp
   please follow the plr for usage
2. Refactor code to catch upstream
3. enhance bytea support in R clients

### CODE REVIEW

File: rcall.c [hunk 1]
```diff
@@ -906,11 +941,299 @@ SEXP plr_SPI_exec(SEXP rsql) {
...
+    }
+
+    msg.msgtype   = MT_SQL;
+    msg.sqltype   = SQL_TYPE_PREPARE;
+	msg.nargs     = nargs;
+    msg.statement = strdup(query);
+	msg.args      = malloc(msg.nargs * sizeof(plcArgument));
+
+    for (i = 0; i < nargs; i++) {
+        char typeid[12];
```
Comment (zhangh43) [inline rcall.c]: magic number

**Comment by zhangh43** [comment]
+1

File: rcall.c [hunk 2]
```diff
@@ -764,10 +781,21 @@ static void pg_get_one_r(char *value,  plcDatatype column_type, SEXP *obj, int e
                            plc_get_type_name(column_type), column_type);
             break;
 
+        case PLC_DATA_BYTEA:
         case PLC_DATA_TEXT:
         default:
             /* Everything else is defaulted to string */
-            SET_STRING_ELT(*obj, elnum, COPY_TO_USER_STRING(((char *)value)));
+			if (strlen(value) != 0 && column_type == PLC_DATA_BYTEA){
```
Comment (paul-guo-) [inline rcall.c]: Move the code for PLC_DATA_BYTEA to the PLC_DATA_BYTEA branch.
Use (value[0] != '\0') instead.

File: rcall.c [hunk 3]
```diff
@@ -764,10 +781,21 @@ static void pg_get_one_r(char *value,  plcDatatype column_type, SEXP *obj, int e
...
+				/*for bytea type, we first get its size then do copy*/
+		    	bsize = *((int*)value);
+				temp = (char *)pmalloc(bsize + 1);
+				strncpy (temp, value + 4, bsize);
+				temp[bsize] = '\0';
+				SET_STRING_ELT(*obj, elnum, COPY_TO_USER_STRING(temp));
```
Comment (paul-guo-) [inline rcall.c]: If bytea allows '\0' before the end, then the code is buggy. Please check, either fix or disable the type.

File: rcall.c [hunk 4]
```diff
@@ -864,8 +892,15 @@ SEXP plr_SPI_exec(SEXP rsql) {
          */
         SET_STRING_ELT(names, j, Rf_mkChar(result->names[j]));
 
-        //create a vector of the type that is rows long
-        PROTECT(fldvec = get_r_vector(result->types[0].type, result->rows));
+        /* 
+		 * create a vector of the type that is rows long
+		 * For type BYTEA, we process it as TEXT
+		 */
```
Comment (paul-guo-) [inline rcall.c]: Format.

File: rcall.c [hunk 5]
```diff
@@ -906,11 +941,299 @@ SEXP plr_SPI_exec(SEXP rsql) {
...
-void raise_execution_error (const char *format, ...) {
-    char *msg   = NULL;
+/*
+ * plr_SPI_prepare - The builtin SPI_prepare command for the R interpreter
+ */
+SEXP plr_SPI_prepare(SEXP rsql, SEXP rargtypes)
+{
+	const char		   *query;
+	int					nargs;
+    int                 res;
```
Comment (paul-guo-) [inline rcall.c]: format

File: rcall.c [hunk 6]
```diff
@@ -906,11 +941,299 @@ SEXP plr_SPI_exec(SEXP rsql) {
...
+
+	lprintf(NOTICE, "start to prepare");
+    if (query == NULL) {
+        raise_execution_error("cannot prepare empty query");
+        return NULL;
+    }
+
+	PROTECT(rargtypes = AS_INTEGER(rargtypes));
+	if (!isVector(rargtypes) || !isInteger(rargtypes)) {
+        raise_execution_error("second parameter must be a vector of PostgreSQL datatypes");
```
Comment (paul-guo-) [inline rcall.c]: Need to return for all raise_execution_error().

File: rcall.c [hunk 7]
```diff
@@ -906,11 +941,299 @@ SEXP plr_SPI_exec(SEXP rsql) {
...
+    }
+
+	/* deal with case of no parameters for the prepared query */
+	if (rargtypes == R_MissingArg || INTEGER(rargtypes)[0] == NA_INTEGER) {
+        nargs = 0;
+    } else {
+        nargs = length(rargtypes);
+    }
+
+	if (nargs < 0) {
```
Comment (paul-guo-) [inline rcall.c]: unprotect rargtypes here.

File: rcall.c [hunk 8]
```diff
@@ -906,11 +941,299 @@ SEXP plr_SPI_exec(SEXP rsql) {
...
+        raise_execution_error("second parameter must be a vector of PostgreSQL datatypes");
+    }
+
+    msg.msgtype   = MT_SQL;
+    msg.sqltype   = SQL_TYPE_PREPARE;
+	msg.nargs     = nargs;
+    msg.statement = strdup(query);
+	msg.args      = malloc(msg.nargs * sizeof(plcArgument));
```
Comment (paul-guo-) [inline rcall.c]: Format. Please check the whole patch. Do not mix tab and space.

File: rcall.c [hunk 9]
```diff
@@ -906,11 +941,299 @@ SEXP plr_SPI_exec(SEXP rsql) {
...
+    plcontainer_channel_send(conn, (plcMessage*) &msg);
+	free_arguments(args, nargs, false, false);
+
+receive:
+    res = plcontainer_channel_receive(plcconn_global, &resp);
+    if (res < 0) {
+        raise_execution_error("Error receiving data from the backend, %d", res);
+        return NULL;
+    }
+
```
Comment (paul-guo-) [inline rcall.c]: You could extract the duplicated code into one function lik that in python. See receive_from_frontend()

File: rconversions.c [hunk 10]
```diff
@@ -193,6 +192,7 @@ static SEXP plc_r_object_from_array (char *input, plcRType *type) {
                  case PLC_DATA_TEXT:
                  default:
                      /* Everything else is defaulted to string */
+					lprintf(NOTICE, "try convert from array, data is %s:[%d]", plc_get_type_name(arr->meta->type), arr->meta->type);
```
Comment (paul-guo-) [inline rconversions.c]: Why NOTICE here?

**Pull request merged**